### PR TITLE
cooling: PolicyX files removed

### PR DESCRIPTION
**Pull request by relh**

### CODE REVIEW

**Comment by relh** [comment]
Training locally, training on a mettabox, and training on distributed works. 

Reloading works locally, on a mettabox, and on distributed multi-gpu training (haven't checked multi-node yet)

image

File: agent/src/metta/agent/metta_agent.py [hunk 1]
```diff
@@ -302,118 +283,5 @@ def _convert_logit_index_to_action(self, logit_indices: torch.Tensor) -> torch.T
         # Default implementation using MettaAgent's action tensors
         return self.action_index_tensor[logit_indices]
 
-    def __setstate__(self, state):
```
Comment (nishu-builder) [inline agent/src/metta/agent/metta_agent.py:305]: I only see this used in tests, so I assume it's safe to remove?

File: agent/src/metta/agent/metta_agent.py [hunk 2]
```diff
@@ -247,12 +224,16 @@ def _apply_feature_remapping(self, features: dict[str, dict], unknown_id: int):
             if feature_id not in self.feature_id_remap and feature_id not in current_feature_ids:
                 remap_tensor[feature_id] = unknown_id
 
-        self.policy._apply_feature_remapping(remap_tensor)
+        # Apply remapping to policy if it supports it
+        if hasattr(self.policy, "_apply_feature_remapping"):
```
Comment (nishu-builder) [inline agent/src/metta/agent/metta_agent.py:228]: i see that even PytorchAgentMixin has _apply_feature_remapping defined; do we need the hasattr check?

File: agent/src/metta/agent/metta_agent.py [hunk 3]
```diff
@@ -247,12 +224,16 @@ def _apply_feature_remapping(self, features: dict[str, dict], unknown_id: int):
...
+            self.policy._apply_feature_remapping(remap_tensor)
+
         self._update_normalization_factors(features)
 
     def _update_normalization_factors(self, features: dict[str, dict]):
         """Update normalization factors after feature remapping."""
-        self.policy.update_normalization_factors(features, getattr(self, "original_feature_mapping", None))
+        if hasattr(self.policy, "update_normalization_factors"):
```
Comment (nishu-builder) [inline agent/src/metta/agent/metta_agent.py:235]: this is only ever called by _apply_feature_remapping so im not sure we need he hasattr check

File: metta/core/distributed.py [hunk 4]
```diff
@@ -29,15 +29,18 @@ def setup_torch_distributed(device: str) -> TorchDistributedConfig:
     distributed = False
 
     if "LOCAL_RANK" in os.environ and device.startswith("cuda"):
+        local_rank = int(os.environ["LOCAL_RANK"])
         torch.distributed.init_process_group(backend="nccl")
 
-        torch.cuda.set_device(device)
+        # Set device using local_rank integer to avoid parsing issues
+        torch.cuda.set_device(local_rank)
```
Comment (nishu-builder) [inline metta/core/distributed.py:36]: deferring to you on this; device would be e.g. "cuda:{local_rank}" so we are passing something different into set_device than before

File: metta/eval/eval_service.py [hunk 5]
```diff
@@ -28,30 +27,36 @@ def evaluate_policy(
...
     stats_dir = stats_dir or "/tmp/stats"
 
-    # For each checkpoint of the policy, simulate
-    logger.info(f"Evaluating policy {pr.uri}")
+    logger.info(f"Evaluating checkpoint {checkpoint_uri}")
     if not is_unique([sim.name for sim in simulations]):
         raise ValueError("Simulation names must be unique")
 
+    # Load the policy from URI
+    policy = CheckpointManager.load_from_uri(checkpoint_uri)
```
Comment (nishu-builder) [inline metta/eval/eval_service.py:41]: for the future: load_from_uri seems like it could support `to_device` or something? i could see forgetting to do the equivalent of lines 44-45 being a common mistake

**Review by nishu-builder — CHANGES_REQUESTED** [review]

File: metta/core/distributed.py (hunk 4 above)
Comment (relh) [reply metta/core/distributed.py:36]: yeah I was mucking about with distributed friday afternoon.. gonna fix it now

File: agent/src/metta/agent/metta_agent.py (hunk 1 above)
Comment (relh) [reply agent/src/metta/agent/metta_agent.py:305]: yeah this function is called only when pytorch models are reloaded from disk

File: metta/rl/checkpoint_manager.py [hunk 6]
```diff
@@ -1,302 +1,380 @@
...
+        """Convert path to URI format and expand short wandb URIs."""
+        if not path_or_uri.startswith(("file://", "wandb://", "s3://", "mock://")):
+            return f"file://{Path(path_or_uri).resolve()}"
+        if path_or_uri.startswith("wandb://"):
+            return expand_wandb_uri(path_or_uri)
+        return path_or_uri
+
+
+    @staticmethod
+    def get_policy_metadata(uri: str) -> dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (nishu-builder) [reply metta/rl/checkpoint_manager.py:226]: i understand it's against everything you stand for so i will understand if you opt not to

File: metta/rl/checkpoint_manager.py [hunk 7]
```diff
@@ -1,302 +1,380 @@
...
+        return _parse_uri_path(uri, "mock"), 0
+
+    return "unknown", 0
+
+
+def is_valid_checkpoint_filename(filename: str) -> bool:
+    """Check if filename matches expected checkpoint format."""
+    if not filename.endswith(".pt"):
+        return False
+    parts = filename[:-3].split("__")
```
Comment (nishu-builder) [inline metta/rl/checkpoint_manager.py:94]: we may want to add something in TrainTool that checks that run_name doesnt use this string anywhere

File: metta/rl/checkpoint_manager.py [hunk 8]
```diff
@@ -1,302 +1,380 @@
...
-        logger.warning(f"Saved policy record did not have a uri: {new_record}")
+def parse_checkpoint_filename(filename: str) -> tuple[str, int, int, int, float]:
+    """Parse checkpoint metadata from filename."""
+    if not is_valid_checkpoint_filename(filename):
+        raise ValueError(f"Invalid checkpoint filename format: {filename}")
+    parts = filename[:-3].split("__")
+    return (parts[0], int(parts[1][1:]), int(parts[2][1:]), int(parts[3][1:]), int(parts[4][2:]) / 10000.0)
+
+
+def _find_best_checkpoint_in_dir(directory: Path) -> Optional[Path]:
```
Comment (nishu-builder) [inline metta/rl/checkpoint_manager.py:116]: name implies it's looking at a score; maybe find_latest_?

File: metta/rl/checkpoint_manager.py [hunk 9]
```diff
@@ -1,302 +1,380 @@
...
+    """Find the best checkpoint file in a directory."""
+    # Try direct directory first, then checkpoints subdirectory
+    search_dirs = [directory]
+    if directory.name != "checkpoints":
```
Comment (nishu-builder) [inline metta/rl/checkpoint_manager.py:120]: are we ever not passing in the checkpoint dir directly?

File: metta/rl/checkpoint_manager.py [hunk 10]
```diff
@@ -1,302 +1,380 @@
...
+        """Load a policy from file://, s3://, or wandb:// URI.
+
+        Supports loading from local files, S3 buckets, wandb artifacts, or mock URIs.
+        Defaults to CPU if no device is specified. Returns None if loading fails.
+        """
+        # Normalize the URI first (converts plain paths to file:// URIs)
+        uri = CheckpointManager.normalize_uri(uri)
+
+        # Convert device to string for consistency
+        if device is None:
```
Comment (nishu-builder) [inline metta/rl/checkpoint_manager.py:173]: looks like torch.load supports supports str or torch.device, so we dont need to turn this into device_str i think

File: metta/rl/trainer.py [hunk 11]
```diff
@@ -527,37 +562,46 @@ def train(
...
-                            device=device,
-                            vectorization=system_cfg.vectorization,
-                            replay_dir=trainer_cfg.evaluation.replay_dir,
-                            stats_epoch_id=stats_tracker.stats_epoch_id,
-                            wandb_policy_name=wandb_policy_name,
-                            policy_store=policy_store,
-                            stats_client=stats_client,
-                            logger=logger,
-                        )
+                        logger.warning("Local evaluation not configured")
```
Comment (nishu-builder) [inline metta/rl/trainer.py:595]: david's suggested we do want to allow for local evaluation still, and I agree --
- non-softmax people
- if you just want to see some immediate results
- if youre changing the evaluation code itself

Comment (nishu-builder) [reply metta/rl/trainer.py:595]: sadly i think this means some of your refactoring for the removal of rl/stats may need to get unwound :(

File: metta/setup/local_commands.py [hunk 12]
```diff
@@ -97,21 +95,7 @@ def main(self, argv=None) -> None:
         elif args.command == "build-app-backend-img":
             self.build_app_backend_img()
         elif args.command == "load-policies":
-            # Convert back to list format for compatibility
```
Comment (nishu-builder) [inline metta/setup/local_commands.py:100]: feel free to just remove load-policies  entirely, nobody uses it except me i think

File: metta/setup/shell.py [hunk 13]
```diff
@@ -36,6 +20,11 @@ def help_configs() -> None:
     info('cfg = load_cfg("trainer/trainer.yaml")')
     success("# Load configs with overrides:")
     info('cfg = load_cfg("train_job.yaml", ["trainer.curriculum=/env/mettagrid/arena/advanced"])')
+    success("# Load checkpoints:")
+    info('policy = CheckpointManager.load_from_uri("file://./train_dir/my_run/checkpoints")')
```
Comment (nishu-builder) [inline metta/setup/shell.py:24]: nice ty

File: metta/sim/simulation.py [hunk 14]
```diff
@@ -174,8 +165,13 @@ def create(
         run_name: str = "simulation_run",
     ) -> "Simulation":
         """Create a Simulation with sensible defaults."""
-        # Get policy record or create a mock
-        policy_record = policy_store.policy_record_or_mock(policy_uri, run_name)
+        # Create policy record from URI
+        if policy_uri:
```
Comment (nishu-builder) [inline metta/sim/simulation.py:169]: we probably want to fail loudly if your policy isnt found?

File: agent/src/metta/agent/pytorch/fast.py [hunk 15]
```diff
@@ -95,16 +93,12 @@ def __init__(self, env, input_size=128, hidden_size=128):
         self.is_continuous = False
         self.action_space = env.single_action_space
 
-        self.out_width = env.obs_width if hasattr(env, "obs_width") else 11
-        self.out_height = env.obs_height if hasattr(env, "obs_height") else 11
+        self.out_width = env.obs_width
```
[the remark this answers is not in the document]
Comment (relh) [reply agent/src/metta/agent/pytorch/fast.py:96]: yeah, this is on purpose. clearly the downstream result of a hasattr removal but I know fast.py so it's fine.

[the remark this answers is not in the document]
Comment (relh) [reply agent/src/metta/agent/pytorch/fast.py:96]: the reason to even have self.out_width as a member variable is that it was supposed to be used later and make readability easier

**Review by nishu-builder — APPROVED** [review]
